src/util/util_env.cpp

```diff
@@ -68,3 +53,23 @@ namespace dxvk::env {
   }
   
 }
+
+#ifdef CP_UNIXCP
+static constexpr int cp = CP_UNIXCP;
+#else
+static constexpr int cp = CP_ACP;
+#endif
+
+namespace dxvk::str {
+    std::string fromws(const WCHAR *ws) {
+        size_t len = WideCharToMultiByte(cp, 0, ws, -1, NULL, 0, NULL, NULL);
+        if (len <= 1)
+            return "";
+
+        len--;
+        std::string result;
+        result.resize(len);
+        WideCharToMultiByte(cp, 0, ws, -1, &result.at(0), len, NULL, NULL);
```

Comment:
Will this work with things like CJK file names when using either the mingw build or the winelib build?

Reply:
I admit that I didn't test it too extensively, but I'd expect it to work (to the degree that Windows allows it using non-unicode APIs). I may do more testing, but here is some more background:

- on mingw build
When you do std::ofstream(str, ...), standard c++ library will end up calling one of msvcrt.*open variants, passing your string argument. Windows internally uses utf-16 for things like file names, so it needs to be converted to utf-16 at some point. In this case, msvcrt will then convert the string to urf-16 using default system codepage (CP_ACP) and pass that to lower level functions (CreateFileW). So if A->utf16 uses CP_ACP, utf16->A should also use CP_ACP for things to work reliably.

Obviously, current codepage may or may not support CJK. But if it doesn't, there is no way we may support that using ansi-based APIs. This is general limitation of the way those APIs work on Windows (and Wine needs to be compatible with Windows in this aspect). Passing utf8 string will not help in this case, because Windows will interpret that as different codepage anyway (well, unless utf8 is the system codepage).

The only way to avoid the problem is to avoid all conversions. It would require using utf16 APIs like kernel32 CreateFileW or msvcrt _wopen. There is no portable C++ way of doing that. MSVC C++ library allows passing wchar_t strings to constructors and functions, but GCC stdc++ does not. There are some tricks like what Mozilla does:
https://hg.mozilla.org/mozilla-central/file/tip/mfbt/FStream.h#l32
but I don't think that's worth it in case of dxvk.

- on winelib build
std::ofstream just ends up calling glibc's open. There are no Windows paths nor Windows encoding involved. It's just plain UNIX path name. Linux treats file names just as byte strings and does not attempt to do any conversions or decoding. It's standard practice that those byte strings are interpreted using configured encoding and that's what Wine's CP_UNIXCP extension uses. Those days it probably means utf8 for majority users, but it depends on configuration.